File: backend/src/database/wallet.rs
use anyhow::{Context, Ok, Result};
use chrono::Utc;
// use tracing_subscriber::fmt::time::FormatTime;

// use crate::api::wallet;

use super::{DbPool, Wallet};

// Insert new wallet into database
pub async fn create_wallet(pool: &DbPool, wallet_name: &str, wallet_id: &str) -> Result<Wallet> {
    let now = Utc::now().to_rfc3339();

    let wallet = sqlx::query_as!(
        Wallet,
        r#"
        INSERT INTO wallets (name, wallet_id, created_at, updated_at)
        VALUES (?, ?, ?, ?)
        RETURNING id, name, wallet_id, created_at, updated_at
        "#,
        wallet_name,
        wallet_id,
        now,
        now,
    )
    .fetch_one(pool)
    .await
    .context("Failed to create wallet in database")?;

    Ok(wallet)
}

// Find wallet by UUID
pub async fn get_wallet_by_id(pool: &DbPool, wallet_id: &str) -> Result<Option<Wallet>> {
    let wallet = sqlx::query_as!(
        Wallet,
        r#"
        SELECT id, name, wallet_id, created_at, updated_at
        FROM wallets 
        WHERE wallet_id = ?
        "#,
        wallet_id,
    )
    .fetch_optional(pool)
    .await
    .context("Failed to query wallet in database")?;

    Ok(wallet)
}

// // Get all wallets (for future list endpoint)
pub async fn list_wallets(pool: &DbPool) -> Result<Vec<Wallet>> {
    let wallets = sqlx::query_as!(
        Wallet,
        r#"
        SELECT *
        FROM wallets
        "#,
    )
    .fetch_all(pool) // for multiple results
    .await
    .context("Failed to query list of wallets")?;

    Ok(wallets)
}

pub async fn update_wallet_name(
    pool: &DbPool,
    wallet_name: &str,
    wallet_id: &str,
) -> Result<Option<Wallet>> {
    let now = Utc::now().to_rfc3339();

    let wallet = sqlx::query_as!(
        Wallet,
        r#"
        UPDATE wallets
        SET name = ?, updated_at = ?
        WHERE wallet_id = ?
        RETURNING id, name, wallet_id, created_at, updated_at
        "#,
        wallet_name,
        now,
        wallet_id,
    )
    .fetch_optional(pool)
    .await
    .context("Failed to update wallet")?;

    Ok(wallet)
}

/// Delete wallet by UUID (DANGEROUS - use with caution)
///
/// This permanently removes wallet metadata from database.
/// In production, consider soft-delete or archive instead.
pub async fn delete_wallet(pool: &DbPool, wallet_id: &str) -> Result<bool> {
    let result = sqlx::query!("DELETE FROM wallets WHERE wallet_id = ?", wallet_id)
        .execute(pool)
        .await
        .context("Failed to delete wallet from database")?;

    // Returns true if wallet was actually deleted
    Ok(result.rows_affected() > 0)
}

// Implement later
// Archive wallet instead of permanent deletion
// pub async fn archive_wallet(pool: &DbPool, wallet_id: &str) -> Result<Option<Wallet>> {
//     let wallet = sqlx::query_as!(
//         Wallet,
//         r#"
//           UPDATE wallets
//           SET updated_at = ?, archived_at = ?
//           WHERE wallet_id = ? AND archived_at IS NULL
//           RETURNING id, name, wallet_id, created_at, updated_at
//         "#,
//         Utc::now().to_rfc3339(),
//         Utc::now().to_rfc3339(),
//         wallet_id,
//     )
//     .fetch_optional(pool)
//     .await
//     .context("Failed to archive wallet")?;

//     Ok(wallet)
// }

File: go-backend/internal/database/wallet.go
package database

import (
	"database/sql"
	"fmt"
	"time"
)

// CreateWallet inserts a new wallet and returns it.
func CreateWallet(db *sql.DB, name, walletID string) (Wallet, error) {
	now := time.Now().UTC().Format(time.RFC3339)

	row := db.QueryRow(
		`INSERT INTO wallets (name, wallet_id, created_at, updated_at)
		 VALUES (?, ?, ?, ?)
		 RETURNING id, name, wallet_id, created_at, updated_at`,
		name, walletID, now, now,
	)

	return scanWallet(row)
}

// GetWalletByID finds a wallet by its UUID. Returns nil if not found.
func GetWalletByID(db *sql.DB, walletID string) (*Wallet, error) {
	row := db.QueryRow(
		`SELECT id, name, wallet_id, created_at, updated_at
		 FROM wallets
		 WHERE wallet_id = ?`,
		walletID,
	)

	w, err := scanWallet(row)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &w, nil
}

// ListWallets returns all wallets.
func ListWallets(db *sql.DB) ([]Wallet, error) {
	rows, err := db.Query("SELECT id, name, wallet_id, created_at, updated_at FROM wallets")
	if err != nil {
		return nil, fmt.Errorf("query wallets: %w", err)
	}
	defer rows.Close()

	var wallets []Wallet
	for rows.Next() {
		w, err := scanWalletRows(rows)
		if err != nil {
			return nil, err
		}
		wallets = append(wallets, w)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate wallets: %w", err)
	}

	if wallets == nil {
		wallets = []Wallet{}
	}
	return wallets, nil
}

// UpdateWalletName updates a wallet's name. Returns nil if not found.
func UpdateWalletName(db *sql.DB, walletID, name string) (*Wallet, error) {
	now := time.Now().UTC().Format(time.RFC3339)

	row := db.QueryRow(
		`UPDATE wallets
		 SET name = ?, updated_at = ?
		 WHERE wallet_id = ?
		 RETURNING id, name, wallet_id, created_at, updated_at`,
		name, now, walletID,
	)

	w, err := scanWallet(row)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &w, nil
}

// DeleteWallet deletes a wallet by its UUID. Returns true if a row was deleted.
// Associated addresses are removed via ON DELETE CASCADE.
func DeleteWallet(db *sql.DB, walletID string) (bool, error) {
	result, err := db.Exec("DELETE FROM wallets WHERE wallet_id = ?", walletID)
	if err != nil {
		return false, fmt.Errorf("delete wallet: %w", err)
	}

	affected, err := result.RowsAffected()
	if err != nil {
		return false, fmt.Errorf("rows affected: %w", err)
	}
	return affected > 0, nil
}

// scanWallet scans a single row into a Wallet.
func scanWallet(row *sql.Row) (Wallet, error) {
	var w Wallet
	var id sql.NullInt64
	var updatedAt sql.NullString

	err := row.Scan(&id, &w.Name, &w.WalletID, &w.CreatedAt, &updatedAt)
	if err != nil {
		return Wallet{}, err
	}

	if id.Valid {
		w.ID = &id.Int64
	}
	if updatedAt.Valid {
		w.UpdatedAt = &updatedAt.String
	}
	return w, nil
}

// scanWalletRows scans from sql.Rows into a Wallet.
func scanWalletRows(rows *sql.Rows) (Wallet, error) {
	var w Wallet
	var id sql.NullInt64
	var updatedAt sql.NullString

	err := rows.Scan(&id, &w.Name, &w.WalletID, &w.CreatedAt, &updatedAt)
	if err != nil {
		return Wallet{}, fmt.Errorf("scan wallet: %w", err)
	}

	if id.Valid {
		w.ID = &id.Int64
	}
	if updatedAt.Valid {
		w.UpdatedAt = &updatedAt.String
	}
	return w, nil
}
